src/System.Text.Http.Parser/HttpParser.cs

```diff
@@ -440,52 +462,73 @@ public HttpParser(bool showErrorDetails)
                     }
                 }
 
-                if (rest != null) {
-                    span = rest.First.Span;
-                    rest = rest.Rest;
-                    remaining = span.Length + remaining;
-                    index = span.Length - remaining;
+                // This is just to get the position to be at the second segment
+                if (position == Position.First)
+                {
+                    if (!buffer.TryGet(ref position, out ReadOnlyMemory<byte> current, advance: true))
+                    {
+                        consumedBytes = 0;
+                        return false;
+                    }
+                }
+
+                if (buffer.TryGet(ref position, out var nextSegment, advance: true))
+                {
+                    currentSpan = nextSegment.Span;
+                    remaining = currentSpan.Length + remaining;
+                    index = currentSpan.Length - remaining;
                 }
-                else {
+                else
+                {
                     consumedBytes = 0;
                     return false;
                 }
             }
         }
 
-        private static unsafe void ReadNextTwoChars(byte* pBuffer, int remaining, int index, out int ch1, out int ch2, IReadOnlyBufferList<byte> rest)
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static void ReadTwoChars(ReadOnlyBytes buffer, int consumedBytes, out int ch1, out int ch2)
         {
-            if (remaining == 1) {
-                ch1 = pBuffer[index];
-                if (rest == null) {
-                    ch2 = -1;
-                }
-                else {
-                    ch2 = rest.First.Span[0];
-                }
-            }
-            else {
+            Span<byte> temp = stackalloc byte[2];
+            if (buffer.Slice(consumedBytes).CopyTo(temp) < 2)
+            {
                 ch1 = -1;
                 ch2 = -1;
-                if (rest != null) {
-                    var nextSpan = rest.First.Span;
-                    if (nextSpan.Length > 0) {
-                        ch1 = nextSpan[0];
-
-                        if (nextSpan.Length > 1) {
-                            ch2 = nextSpan[1];
-                        }
-                        else {
-                            ch2 = -1;
-                        }
-                    }
-                    else {
-                        ch2 = -1;
-                    }
-                }
+            }
+            else
+            {
+                ch1 = temp[0];
+                ch2 = temp[1];
             }
         }
 
+        // This is not needed, but I will leave it here for now, as we might need it later when optimizing
+        //private static unsafe void ReadNextTwoChars(byte* pBuffer, int remaining, int index, out int ch1, out int ch2, ReadOnlySpan<byte> next)
```

Comment:
This is different from the original ReadNextTwoChars. Are the changes intentional?

Reply:
Yeah, all the changes are intentional. I first changed this routine, then simplified at the callsite.